### PR TITLE
Telepath - Release notes and documentation on customisation / porting

### PR DESCRIPTION
**Pull request by gasman**
Based on the StreamField doc reorganisation started at the documentation sprint and completed in #6771, add the necessary API docs and extend the customisation howto page so that anyone with heavily customised blocks under the old implementation now hopefully has enough guidance to port them to the telepath implementation. This finally gives us everything we need to write up the release notes without leaving any gaps...

### CODE REVIEW

File: docs/reference/streamfield_reference.rst [hunk 1]
```diff
@@ -49,472 +49,442 @@ All block types accept the following optional keyword arguments:
...
+   A block consisting of many sub-blocks, all of the same type. The editor can add an unlimited number of sub-blocks, and re-order and delete them. Takes the definition of the sub-block as its first argument:
 
-.. code-block:: python
-    :emphasize-lines: 6
+   .. code-block:: python
 
-    class IngredientsListBlock(blocks.ListBlock):
-        ingredient = blocks.CharBlock()
-        amount = blocks.CharBlock(required=False)
+       ('ingredients_list', blocks.ListBlock(blocks.CharBlock(label="Ingredient")))
```
Comment (kaedroho) [inline docs/reference/streamfield_reference.rst]: If we're going to use the tuple format as an example, could we keep the ``body = StreamField([ ... ])`` around it?

File: docs/reference/streamfield/widget_api.rst [hunk 2]
```diff
@@ -0,0 +1,39 @@
+.. _streamfield_widget_api:
+
+Form widget client-side API
+===========================
+
+In order for the StreamField editing interface to dynamically create form fields, any Django form widgets used within StreamField blocks must have an accompanying JavaScript implementation, defining how the widget is rendered client-side and populated with data, and how to extract data from that field. Wagtail provides this implementation for widgets inheriting from ``django.forms.widgets.Input``, ``django.forms.Textarea``, ``django.forms.Select`` and ``django.forms.RadioSelect``. For any other widget types, or ones which require custom client-side behaviour, you will need to provide your own implementation.
+
+The `telepath <https://wagtail.github.io/telepath/>`__ library is used to set up mappings between Python widget classes and their corresponding JavaScript implementations. To create a mapping, define a subclass of ``wagtail.core.telepath.Adapter`` and register it with ``wagtail.core.telepath.register``.
```
Comment (kaedroho) [inline docs/reference/streamfield/widget_api.rst:8]: I think an example would be useful here

**Review by kaedroho — APPROVED** [review]
LGTM. Nice work!

**Comment by gasman** [comment]
Thanks @kaedroho! Merged with feedback applied.

**Pull request merged**